crates/anvil/src/eth/backend/mem/mod.rs

```diff
@@ -686,6 +693,8 @@ impl Backend {
                 ..Default::default()
             };
         }
+
+        self.active_snapshots.lock().retain(|key, _| *key <= id);
```

Comment:
hmm. so snapshots `> id` are still useful?
I was under the impression they could no longer be used. but perhaps it was just a limitation on anvil's side? never tested that with forge

Reply:
Correct, they're not deleted and can be reused now. To be more specific, snapshots now:
- Can be used repeatedly (e.g, create a snapshot outside a loop, and on every loop iteration, modify state and then revert back to the snapshot created outside the loop)
- Deleting a snapshot does not delete future snapshots. (e.g, deleting snapshot `id` does not delete `id + 1, id + 2... id + n`)

Reply:
sorry, but I'm a bit confused right now

the current behaviour seems to be that snapshots `> id` are no longer usable. at least with anvil:
```
~/contrib/foundry-rs
❯ anvil --version
anvil 0.2.0 (6432031 2023-11-27T00:17:45.338056018Z)

~/contrib/foundry-rs
❯ cast rpc --rpc-url localhost:8545 evm_snapshot
"0x0"

~/contrib/foundry-rs
❯ cast rpc --rpc-url localhost:8545 evm_mine
"0x0"

~/contrib/foundry-rs
❯ cast rpc --rpc-url localhost:8545 evm_snapshot
"0x1"

~/contrib/foundry-rs
❯ cast rpc --rpc-url localhost:8545 evm_revert 0x0
true

~/contrib/foundry-rs
❯ cast rpc --rpc-url localhost:8545 evm_revert 0x1
Error:
(code: -32001, message: Resource not found, data: None)
```

The issues linked seem to be about forge tests, and your fix above is on `evm/`. so perhaps this is just some inconsistency between forge & anvil?

From glancing at the anvil code, I'd guess future snapshots couldn't possibly work, due to [this call](https://github.com/foundry-rs/foundry/blob/f0166ccf0c8dc3ff626eed3c0a060804d8b94cb1/crates/anvil/src/eth/backend/mem/mod.rs#L670) (which seems to be where the error above comes from), but not sure

Reply:
yep! sorry for the confusion here. It's true that anvil and forge behave differently—anvil should match forge's behavior which is what i described above.

Reply:
ok, so it's a bug on anvil's side. I'll try to check what's causing anvil to crash then

Reply:
ok, looking at the code, I don't see how that's achievable with the current logic.
`revert_snapshot` seems to be removing all future data from `self.storage`, but then relies on that same storage to get the target of the snapshot.

If we're reverting into a block that no longer exists under the current state, this fails

Reply:
thanks for taking the time!

it sounds, from what you're describing, that it's a much more complex problem to keep future snapshots around with a node.

seems like it could be warranted to have different behaviour between anvil & forge then (which I believe it's what this PR achieves already, but need to double-check, don't fully remember right now).


as for the decision of whether or not to delete the target snapshot:
The OG behaviour seems a bit weird to me (as in, it's limiting usability for no apparent reason). I'd be fine with break it, but also understand that keeping within spec is important and I probably don't know the full repercussions.

If we do keep it, I can work around that on my app's side: by automatically creating a new snapshot everytime a revert occurs. not ideal, but I can make it work

so I guess it's your call. I can fix the PR according to whatever you decide :)

Reply:
> eems like it could be warranted to have different behaviour between anvil & forge then

agree

> The OG behaviour seems a bit weird to me 

yeah they even mention in the docs that in order to achieve this you're supposed to create a new snapshot after revert.

> I'd be fine with break it

I think that's reasonable.

Though, we'd need additional call or arguments to get that behaviour. because if we keep it then there's currently no way to delete a snapshot.

we can however simply add a new optional bool parameter to evm_revert, this doesn't break the OG behaviour but allows you to force keeping it.

wdyt?

Reply:
> add a new optional bool parameter to evm_revert

didn't think of this, but sounds good!

so, would this be a good plan?
- refactor this PR to simply keep the hardhat behaviour. i.e.: just solve point 3 from the original PR description, which seems to be a bug
- 2nd PR: add the optional `evm_revert(true)` which keeps the target snapshot alive instead of deleting it
- 3rd PR: add `anvil_delete_snapshot(id)` and/or `anvil_clear_snapshots`

Reply:
sg!